libs/js-sdk/src/client.ts

```diff
@@ -110,45 +115,66 @@ export function createBrowserClient({
             })
         },
         subscribeToInitialisedChanged(callback) {
-            initialisedCallbacks.push(callback)
+            debugLog('Subscribing to initialised changed: %o', {
+                initialised: isInitialised(),
+            })
+
+            initialisedState.initialisedCallbacks.push(callback)
             return () => {
-                initialisedCallbacks.splice(
-                    initialisedCallbacks.indexOf(callback),
+                initialisedState.initialisedCallbacks.splice(
+                    initialisedState.initialisedCallbacks.indexOf(callback),
                     1,
                 )
             }
         },
         async updateAudiences(updatedAudiences: string[]) {
+            if (compareArrays(state.audiences, updatedAudiences)) {
+                debugLog('Skipped updating audiences, no change: %o', {
+                    updatedAudiences,
+                    currentAudiences: state.audiences,
+                    initialised: isInitialised(),
+                })
+                // No need to update audiences
+                return Promise.resolve()
+            }
+
             debugLog('Updating audiences: %o', {
                 updatedAudiences,
+                currentAudiences: state.audiences,
                 initialised: isInitialised(),
             })
 
-            if (compareArrays(state.audiences, updatedAudiences)) {
-                // No need to update audiences
-                return Promise.resolve()
-            }
+            const newPromise = new PromiseCompletionSource<boolean>()
+            initialisedState.initialisedPromise = newPromise
+            initialisedState.initialisedPromise.promise.catch(() => {})
+            initialisedState.initialisedPromise.promise.then(() => {
+                // If the promise has changed, then we don't want to invoke the callback
+                if (newPromise !== initialisedState.initialisedPromise) {
+                    return
+                }
 
-            const promise = new PromiseCompletionSource<boolean>()
-            updatedPromise.push(promise)
+                // Get the value from the function, just incase it has changed
+                const initialised = isInitialised()
+                initialisedState.initialisedCallbacks.forEach((c) =>
+                    c(initialised),
+                )
+            })
             debugLog('updateAudiences: invoke initialised callback with false')
-            initialisedCallbacks.forEach((c) => c(false))
+            initialisedState.initialisedCallbacks.forEach((c) => c(false))
 
             state.audiences = updatedAudiences
             debugLog(
                 'updateAudiences: Audiences updated (%o), getting new effective values',
                 updatedAudiences,
             )
 
-            return updateStrategyImplementation.connect(state).then(() => {
-                promise?.resolve(true)
-                if (isInitialised()) {
```

Comment:
This was an issue, because the promise requires a micro tick to change state to completed, so this would always evaluate to false, leaving the SDK in a bad state.